### PR TITLE
Update es5-shim.js String.prototype.trim()

### PR DESCRIPTION
**Pull request by will-v-king**
improve string.trim()
according to:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim
http://stackoverflow.com/questions/2308134/trim-in-javascript-not-working-in-ie/29001502#29001502

### CODE REVIEW

File: es5-shim.js [hunk 1]
```diff
@@ -1408,14 +1408,16 @@ var hasTrimWhitespaceBug = StringPrototype.trim && (ws.trim() || !zeroWidth.trim
 defineProperties(StringPrototype, {
     // http://blog.stevenlevithan.com/archives/faster-trim-javascript
     // http://perfectionkills.com/whitespace-deviations/
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim
     trim: function trim() {
-        if (typeof this === 'undefined' || this === null) {
+        // Make sure we trim BOM and NBSP
+        var rtrim = new RegExp('^'+wsRegexChars+'+|'+wsRegexChars+'+$','g');
+        if ('undefined' === typeof this || null === this) {
```
Comment (ljharb) [inline es5-shim.js]: Please don't switch the order here - I don't like yoda conditions.

**Comment by ljharb** [comment]
Please provide tests that would have failed without this change.

File: es5-shim.js [hunk 2]
```diff
@@ -1408,14 +1408,16 @@ var hasTrimWhitespaceBug = StringPrototype.trim && (ws.trim() || !zeroWidth.trim
 defineProperties(StringPrototype, {
     // http://blog.stevenlevithan.com/archives/faster-trim-javascript
     // http://perfectionkills.com/whitespace-deviations/
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim
     trim: function trim() {
-        if (typeof this === 'undefined' || this === null) {
+        // Make sure we trim BOM and NBSP
+        var rtrim = new RegExp('^'+wsRegexChars+'+|'+wsRegexChars+'+$','g');
```
Comment (ljharb) [inline es5-shim.js]: This regex should be created outside the function, so it's not created every time. See `trimBeginRegexp` and `trimEndRegExp`, which should be removed if they're not used anymore.

**Comment by will-v-king** [comment]
ok, I just test the code in IE console.

**Comment by ljharb** [comment]
That's absolutely not sufficient :-)

Also please see 21c880db which was the last significant modification to this method and make sure the issues in https://gist.github.com/1120592 are covered.

File: es5-shim.js [hunk 3]
```diff
@@ -1402,20 +1402,20 @@ var ws = '\x09\x0A\x0B\x0C\x0D\x20\xA0\u1680\u180E\u2000\u2001\u2002\u2003' +
...
+var rtrim = new RegExp('^'+wsRegexChars+'+|'+wsRegexChars+'+$','g');
 var hasTrimWhitespaceBug = StringPrototype.trim && (ws.trim() || !zeroWidth.trim());
 defineProperties(StringPrototype, {
     // http://blog.stevenlevithan.com/archives/faster-trim-javascript
     // http://perfectionkills.com/whitespace-deviations/
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/trim
     trim: function trim() {
-        if (typeof this === 'undefined' || this === null) {
+        // Make sure we trim BOM and NBSP
+        if (typeof this === 'undefined'  ||this === null) {
```
Comment (ljharb) [inline es5-shim.js]: whitespace?

File: es5-shim.js [hunk 4]
```diff
@@ -1402,20 +1402,20 @@ var ws = '\x09\x0A\x0B\x0C\x0D\x20\xA0\u1680\u180E\u2000\u2001\u2002\u2003' +
...
             throw new TypeError("can't convert " + this + ' to object');
         }
-        return String(this).replace(trimBeginRegexp, '').replace(trimEndRegexp, '');
+        return String(this).replace(rtrim, '');
     }
 }, hasTrimWhitespaceBug);
-
```
Comment (ljharb) [inline es5-shim.js]: Please also revert this line deletion.

File: es5-shim.js [hunk 5]
```diff
@@ -1398,23 +1398,25 @@ defineProperties(StringPrototype, {
 // ES5 15.5.4.20
 // whitespace from: http://es5.github.io/#x15.5.4.20
 var ws = '\x09\x0A\x0B\x0C\x0D\x20\xA0\u1680\u180E\u2000\u2001\u2002\u2003' +
-    '\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028' +
-    '\u2029\uFEFF';
+	'\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028' +
+	'\u2029\uFEFF';
```
Comment (ljharb) [inline es5-shim.js]: Please stick with 4 space indentations, and not hard tabs.

Comment (ljharb) [inline es5-shim.js]: Also, please rebase this back down to a single commit so there's no "formatting" commits :-)

Comment (will-v-king) [inline es5-shim.js]: oh sorry fixing...

**Comment by will-v-king** [comment]
I guess it's ok now?

**Comment by ljharb** [comment]
Now it needs to be rebased down to a single commit, and then tests added.

**Comment by will-v-king** [comment]
And i guess there's no way to re-base in the web.browser, right ?

**Comment by ljharb** [comment]
No, I don't believe so. You'd not be able to run tests in the browser either tho, so you'll have to clone it regardless.

**Comment by will-v-king** [comment]
just clone to local and learning about re-base (I did twice before, but never remembered how to do it.)

**Comment by will-v-king** [comment]
and this is npm test result:
root@lenovo-VirtualBox:/media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim# npm test

> es5-shim@4.1.0 test /media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim
> npm run lint && jasmine-node --matchall ./ tests/spec/


> es5-shim@4.1.0 lint /media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim
> npm run jscs && npm run eslint


> es5-shim@4.1.0 jscs /media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim
> jscs tests/helpers/*.js tests/spec/*.js es5-shim.js es5-sham.js

sh: 1: jscs: not found
npm ERR! weird error 127
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! not ok code 0
npm ERR! weird error 1
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! not ok code 0
npm ERR! weird error 1
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! not ok code 0

The  The Travis CI build is passed

**Comment by will-v-king** [comment]
well since npm install is not succeeded, and i am 1 hour later to go home.  I run the index.html instead.
result is ：

Jasmine1.3.1 revision 1354556913

Show  passed  skipped

run all284 specs, 0 failures in 1.031sFinished at Fri Mar 13 2015 18:08:52 GMT+0800 (中国标准时间)
es5-test

**Comment by will-v-king** [comment]
i run npm install and test on in windows7 and VirtualBox-Ubuntu for several times. Get some ERRs. But finally , do not know how, the test works and passed.
root@lenovo-VirtualBox:/media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim# npm test

> es5-shim@4.1.0 test /media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim
> npm run lint && jasmine-node --matchall ./ tests/spec/


> es5-shim@4.1.0 lint /media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim
> npm run jscs && npm run eslint


> es5-shim@4.1.0 jscs /media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim
> jscs tests/helpers/*.js tests/spec/*.js es5-shim.js es5-sham.js

No code style errors found.

> es5-shim@4.1.0 eslint /media/sf_D_DRIVE/Workspace/2014-09-09_personalTools/jquery-collection/es5-shim
> eslint --reset tests/helpers/*.js tests/spec/*.js es5-shim.js es5-sham.js


tests/spec/s-array.js
  10:18  warning  Missing "use strict" statement  strict

es5-shim.js
   29:8   warning  Missing "use strict" statement         strict
  350:35  warning  deleteCount is defined but never used  no-unused-vars
  591:12  warning  Unexpected constant condition          no-constant-condition
  641:12  warning  Unexpected constant condition          no-constant-condition
  877:44  warning  key is defined but never used          no-unused-vars

es5-sham.js
  29:8  warning  Missing "use strict" statement  strict

✖ 7 problems (0 errors, 7 warnings)

**Comment by will-v-king** [comment]
with JShint in vim I got a message (which sames need to be improved):
es5-shim.js|1310 col 26| Don't make functions within a loop.

**Comment by ljharb** [comment]
That likely means your vim jshint isn't respecting the project's jshintrc file.

**Comment by ljharb** [comment]
So, I think your changes are fine - but, I'd want either a jsperf showing that this is faster, or a *new* test that would have failed without this change.

In other words, without a reason to make the change, I'd prefer not to make it.

**Comment by will-v-king** [comment]
I do not see .jshintrc in this project. 
Whatever, will this pull request be accepted? Or plz let me know what-else I should do to make it happen.

**Comment by ljharb** [comment]
oops, you're right, this project doesn't use `jshint`.

As I said in https://github.com/es-shims/es5-shim/pull/296#issuecomment-81365912, I need either a `jsperf` showing that this is a performance improvement, or, this PR should also contain a *new* test that fails on master.

**Comment by will-v-king** [comment]
Here's a little test script: which proves there's a little improvement on js performance.

tm 20150318120622

var ws = '\x09\x0A\x0B\x0C\x0D\x20\xA0\u1680\u180E\u2000\u2001\u2002\u2003' +
'\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028' +
'\u2029\uFEFF';
var trimBeginRegexp = new RegExp('^' + wsRegexChars + wsRegexChars + '*');
var trimEndRegexp = new RegExp(wsRegexChars + wsRegexChars + '*$');
var rtrim = new RegExp('^' + wsRegexChars + '+|' + wsRegexChars + '+$', 'g');
var zeroWidth = '\u200b';
var wsRegexChars = '[' + ws + ']';

String.prototype.trim1 = function(){
  if (typeof this === 'undefined' || this === null) {
    //throw new TypeError("can't convert " + this + ' to object');
  }
  return String(this).replace(trimBeginRegexp, '').replace(trimEndRegexp, '');
};

String.prototype.trim2 = function(){
  if (typeof this === 'undefined' || this === null) {
    //throw new TypeError("can't convert " + this + ' to object');
  }
  return String(this).replace(rtrim, '');
};

var time = (new Date()).getTime();
var i,count = 1000000;
var exampleStr = " a b c ";
for(i=0; i < count; i++){
    exampleStr.trim1();
}
console.log((new Date()).getTime() - time);
time = (new Date()).getTime();
for(i=0; i < count; i++){
    exampleStr.trim2();
}
console.log((new Date()).getTime() - time);

**Comment by ljharb** [comment]
Please prepare a http://jsperf.com test case; that's typically what's used to demonstrate performance, and facilitate cross-browser testing.

**Comment by will-v-king** [comment]
Here's the test case:
http://jsperf.com/string-trim-performance-test

**Comment by hazzik** [comment]
For me it says that trim2 is slower

Chrome 41.0.2272.89 32-bit on Windows NT 6.3 64-bit

**Comment by will-v-king** [comment]
Well, according to the test result, the original implement is faster.

**Pull request closed without merging**